runtime/runtime-params-estimator/src/trie.rs

```diff
@@ -0,0 +1,413 @@
+    // larger the amortization effect is.
+    num_values: usize,
+    // Values to read before measurement, to warm up CPU caches with data
+    // structures used in looking up trie nodes
+    num_warmup_values: usize,
+    // Spread factor to reduce data locality
+    data_spread_factor: usize,
+    // Before measuring, completely overwrite L3 CPU cache content
+    spoil_l3: bool,
+) -> Vec<GasCost> {
+    // Trie nodes are largest when they hold part of a storage key (Extension or
+    // Leaf) and keys can be up to 2kiB. Therefore, this is about the maximum
+    // size possible.
+    let value_len: usize = 2048;
+
+    // Prepare a data buffer that we can read again later to overwrite CPU caches
+    let assumed_max_l3_size = 128 * 1024 * 1024;
+    let dummy_data = testbed.transaction_builder().random_vec(assumed_max_l3_size);
+
+    (0..iters)
+        .map(|i| {
+            let tb = testbed.transaction_builder();
+            let signer = tb.random_account();
+            let values_inserted = num_values * data_spread_factor;
+            let values: Vec<_> = (0..values_inserted)
+                .map(|_| {
+                    let v = tb.random_vec(value_len);
+                    let h = hash(&v);
+                    let node = RawTrieNode::Extension(v, h);
+                    let node_with_size = RawTrieNodeWithSize { node, memory_usage: 1 };
+                    node_with_size.encode().unwrap()
+                })
+                .collect();
+            let mut setup_block = Vec::new();
+            for (j, value) in values.iter().cloned().enumerate() {
+                let key = j.to_le_bytes().to_vec();
+                setup_block.push(tb.account_insert_key_bytes(signer.clone(), key, value));
+            }
+            testbed.process_block(setup_block, 0);
+
+            // Collect keys of the inserted nodes and select a subset for testing.
+            let all_value_hashes: Vec<_> = values.iter().map(|value| hash(value)).collect();
+            let measured_value_hashes: Vec<_> =
+                all_value_hashes.iter().step_by(data_spread_factor).cloned().collect();
+            let unmeasured_value_hashes = &all_value_hashes[0..num_warmup_values];
+            assert_eq!(measured_value_hashes.len(), num_values);
+
+            // Create a new cache and load nodes into it as preparation.
+            let caching_storage = testbed.trie_caching_storage();
+            caching_storage.set_mode(TrieCacheMode::CachingChunk);
+            let _dummy_sum = read_raw_nodes_from_storage(&caching_storage, &all_value_hashes);
+
+            // Remove trie nodes from CPU caches by filling the caches with useless data.
+            // (To measure latency from main memory, not CPU caches)
+            if spoil_l3 {
+                let dummy_count = dummy_data.iter().filter(|n| **n == i as u8).count();
+                SINK.fetch_add(dummy_count, Ordering::SeqCst);
+            }
```

Comment:
I wonder if we should just have 

```
pub fn spoil_l3_cache() {}
```

utility function?

Or would it actually require `L3Spoiler` struct to cache the random data? 

Not an action item, mostly just thoughts!

Reply:
Yes, I think that would work as utility function! I will consider refactoring, next time I need to spoil L3 in my code.

But, I would use it somewhat sparingly. It can slow down estimations quite a bit. Probably only worth it for costs that are memory latency bound.